packages/hash/frontend/src/pages/[account-slug]/entities/[entity-id].page/entity-editor/properties-section/property-table/use-row-data/generate-property-rows-from-entity/generate-property-row-recursively.ts

```diff
@@ -0,0 +1,103 @@
+import { isPlainObject } from "lodash";
+import {
+  getPropertyTypesByBaseUri,
+  RootEntityAndSubgraph,
+} from "../../../../../../../../../lib/subgraph";
+import { PropertyRow } from "../../types";
+import { getDataTypesOfPropertyType } from "./get-data-types-of-property-type";
+
+/**
+ * This function generates property row data,
+ * and calls itself again for each nested property. Then puts results of these recursive calls into `children` array
+ *
+ * @param properties
+ * Properties object for current depth. On entity level, it starts from `entity.properties`
+ *
+ * @param propertyTypeBaseUri
+ * Name of the specific property inside `properties` object
+ *
+ * @param propertyKeyChain
+ * An array of `propertyTypeBaseUri`'s to store the path to each nested property.
+ * ```js
+ * // Example: `propertyKeyChain` of `c`
+ * properties = { a: { b: { c: "John" } } };
+ * propertyKeyChain = ["a", "b", "c"]
+ * ```
+ *
+ * @param rootEntityAndSubgraph
+ * An object storing root entity & subgraph of that entity
+ *
+ * @param requiredPropertyTypes
+ * An array of `propertyTypeBaseUri`'s.
+ * This is used to check if a property should be shown as `required` or not
+ *
+ * @param depth
+ * Depth of the property. For properties at root, depth starts from `0`
+ *
+ * @returns property row (and nested rows as `children` if it's a nested property)
+ */
+export const generatePropertyRowRecursively = (
+  properties: any,
+  propertyTypeBaseUri: string,
+  propertyKeyChain: string[],
+  rootEntityAndSubgraph: RootEntityAndSubgraph,
+  requiredPropertyTypes: string[],
+  depth: number,
+): PropertyRow => {
+  const propertyTypeVersions = getPropertyTypesByBaseUri(
+    rootEntityAndSubgraph.subgraph,
+    propertyTypeBaseUri,
+  );
+
+  if (!propertyTypeVersions) {
+    throw new Error(
+      `propertyType not found for base URI: ${propertyTypeBaseUri}`,
+    );
+  }
+
+  const propertyType = propertyTypeVersions[0]!.inner;
```

Comment:
@Alfred-Mountfield any idea?

Reply:
I think there's a TODO to think about what we're actually doing here, I'm not sure the TODO is necessarily about some other task.

I think the point here is what do you actually want as the property type here? 

Let's say the entity has a property: `http://hash.ai/@alfie/types/property-type/name`

and lets say the subgraph has the following property types:
* `http://hash.ai/@alfie/types/property-type/name/v/1`
* `http://hash.ai/@alfie/types/property-type/name/v/2`
* `http://hash.ai/@alfie/types/property-type/name/v/3`

Which one do you want to use?

I'm not sure it's safe or correct to just pick the latest version, I _think_ we want to look at the `entity.entityTypeId`, _go to the Entity Type_ (which _should_ be in the subgraph if we made the query correctly), and find the property type in the subgraph.

e.g. if our entity has an `entityTypeId` `http://hash.ai/@alfie/types/entity-type/person/v/2`

we go and look at `person/v/2` and see that it's using `http://hash.ai/@alfie/types/property-type/name/v/1`, so we use that one.

Reply:
Hmm, I don't think we should do this in this PR (even if we want to do this now), because this was not introduced by this PR (even if it looks like that, this PR just moved the existing line to a separete file 🤦‍♂️). But we can capture this in Asana. What do you guys think about this? @nathggns @Alfred-Mountfield

Reply:
There's a similar issue here:
<img width="616" alt="image" src="https://user-images.githubusercontent.com/25749103/199487843-097714dd-d4fe-4277-a4f3-8c7e5940890b.png">

(lib/entities.ts in my branch)

An asana task makes sense to me. Perhaps we can discuss this in the knowledge-share tomorrow

Reply:
As discussed offline @yusufkinatas, I made the task: https://app.asana.com/0/1200211978612931/1203285785211172/f